ecs/pod_creator.go

```diff
@@ -27,11 +31,227 @@ func NewBasicECSPodCreator(c cocoa.ECSClient, v cocoa.Vault) (*BasicECSPodCreato
+
+	if len(runOut.Tasks) == 0 || runOut.Tasks[0].TaskArn == nil {
+		return nil, errors.New("missing running task")
+	}
+
+	resources := cocoa.NewECSPodResources().
+		SetCluster(*mergedPodExecutionOpts.Cluster).
+		SetSecrets(allSecrets).
+		SetTaskDefinition(*taskDef).
+		SetTaskID(*runOut.Tasks[0].TaskArn)
+
+	options := NewBasicECSPodOptions().
+		SetClient(m.client).
+		SetVault(m.vault).
+		SetStatus(cocoa.Running).
+		SetResources(*resources)
+
+	p, err := NewBasicECSPod(options)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating pod")
+	}
+
+	return p, nil
 }
 
 // CreatePodFromExistingDefinition creates a new pod backed by AWS ECS from an
 // existing definition.
 func (m *BasicECSPodCreator) CreatePodFromExistingDefinition(ctx context.Context, def cocoa.ECSTaskDefinition, opts ...*cocoa.ECSPodExecutionOptions) (cocoa.ECSPod, error) {
 	return nil, errors.New("TODO: implement")
 }
+
+// translateStringArrayToECSTagArray translates the strings into ECS tags.
+func translateStringArrayToECSTagArray(tags []string) []*ecs.Tag {
+	ecsTags := []*ecs.Tag{}
+	for _, tag := range tags {
+		ecsTag := &ecs.Tag{}
+		ecsTag.SetKey(tag)
+		ecsTags = append(ecsTags, ecsTag)
+	}
+	return ecsTags
+}
+
+// translateStrategy translates the strategy and parameter into ECS placement strategy.
+func translateStrategy(strategy *cocoa.ECSPlacementStrategy, param *cocoa.ECSStrategyParameter) []*ecs.PlacementStrategy {
+	placementStrat := ecs.PlacementStrategy{}
+	placementStrat.SetType(string(*strategy)).SetField(*param)
+	return []*ecs.PlacementStrategy{&placementStrat}
+}
+
+// exportEnvVars translates the environment variables into ECS environment variables and secrets.
+func (m *BasicECSPodCreator) exportEnvVars(ctx context.Context, variables []cocoa.EnvironmentVariable) ([]*ecs.KeyValuePair, []cocoa.PodSecret, []cocoa.NamedSecret, error) {
+	keyValuePairs := []*ecs.KeyValuePair{}
+	secrets := []cocoa.PodSecret{}
+	createSecrets := []cocoa.NamedSecret{}
+
+	for _, variable := range variables {
+		if variable.SecretOpts == nil {
+			keyValue := ecs.KeyValuePair{}
+			keyValue.SetName(*variable.Name).SetValue(*variable.Value)
+			keyValuePairs = append(keyValuePairs, &keyValue)
+		} else {
+			if !*variable.SecretOpts.Exists {
+				createSecrets = append(createSecrets, variable.SecretOpts.PodSecret.NamedSecret)
+			}
+			secrets = append(secrets, variable.SecretOpts.PodSecret)
+		}
+	}
+
+	return keyValuePairs, secrets, createSecrets, nil
+}
+
+// exportSecrets extracts all secrets and environment variables from ECS creation options
+func (m *BasicECSPodCreator) exportSecrets(ctx context.Context, merged cocoa.ECSPodCreationOptions) ([]*ecs.KeyValuePair, []cocoa.PodSecret, []cocoa.NamedSecret, error) {
+	allEnvVars := []*ecs.KeyValuePair{}
+	allSecrets := []cocoa.PodSecret{}
+	allSecretsToCreate := []cocoa.NamedSecret{}
+
+	for _, def := range merged.ContainerDefinitions {
+		envVars, secrets, secretsToCreate, err := m.exportEnvVars(ctx, def.EnvVars)
```

Comment:
This is probably doing too many things at once. In general, try to keep a function to a single purpose (and follow [these guidelines](https://stackify.com/solid-design-principles/)).

I would separate `exportEnvVars` into two different functions, one to get the unsecret environment variables (`exportEnvVars`, which will return ECS API key-value pairs) and one to extract the secrets, regardless of whether they exist or not (`getSecrets`).

Later, during `createSecrets` you can just check `Exists` to decide if the secret needs to be created or if it can be ignored.